server/src/test/java/org/elasticsearch/discovery/HandshakingTransportAddressConnectorTests.java

```diff
@@ -0,0 +1,178 @@
+            @Override
+            public void onResponse(DiscoveryNode discoveryNode) {
+                receivedNode.set(discoveryNode);
+                completionLatch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError(e);
+            }
+        });
+
+        assertTrue(completionLatch.await(1, TimeUnit.SECONDS));
+        assertEquals(remoteNode, receivedNode.get());
+    }
+
+    public void testDoesNotConnectToNonMasterNode() throws InterruptedException {
+        remoteNode = new DiscoveryNode("remote-node", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT);
+        remoteClusterName = "local-cluster";
+
+        FailureListener failureListener = new FailureListener();
+        handshakingTransportAddressConnector.connectToRemoteMasterNode(remoteNode.getAddress(), failureListener);
+        failureListener.assertFailure();
+    }
+
+    public void testDoesNotConnectToLocalNode() throws Exception {
+        remoteNode = localNode;
+        remoteClusterName = "local-cluster";
+
+        FailureListener failureListener = new FailureListener();
+        handshakingTransportAddressConnector.connectToRemoteMasterNode(remoteNode.getAddress(), failureListener);
+        failureListener.assertFailure();
+    }
+
+    public void testDoesNotConnectToDifferentCluster() throws InterruptedException {
+        remoteNode = new DiscoveryNode("remote-node", buildNewFakeTransportAddress(), Version.CURRENT);
+        remoteClusterName = "another-cluster";
+
+        FailureListener failureListener = new FailureListener();
+        handshakingTransportAddressConnector.connectToRemoteMasterNode(remoteNode.getAddress(), failureListener);
+        failureListener.assertFailure();
+    }
+
+    public void testHandshakeTimesOut() throws InterruptedException {
+        remoteNode = new DiscoveryNode("remote-node", buildNewFakeTransportAddress(), Version.CURRENT);
+        remoteClusterName = "local-cluster";
+        dropHandshake = true;
+
+        FailureListener failureListener = new FailureListener();
+        handshakingTransportAddressConnector.connectToRemoteMasterNode(remoteNode.getAddress(), failureListener);
+        Thread.sleep(PROBE_HANDSHAKE_TIMEOUT_SETTING.get(Settings.EMPTY).millis());
+        failureListener.assertFailure();
+    }
+
+    private class FailureListener implements ActionListener<DiscoveryNode> {
+        final CountDownLatch completionLatch = new CountDownLatch(1);
+
+        @Override
+        public void onResponse(DiscoveryNode discoveryNode) {
+            fail(discoveryNode.toString());
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            completionLatch.countDown();
+        }
+
+        void assertFailure() throws InterruptedException {
+            assertTrue(completionLatch.await(1, TimeUnit.SECONDS));
```

Comment:
maybe this is a little optimistic for our slow CI machines. 30 secs?
Same comment for other places in this class.